### DISCUSSION TITLE
Help, generated Prisma shape mismatches with Prisma Client type

### DISCUSSION THREAD

**Discussion by 2coo** [hayes/pothos#1193] [Q&A] [answered]
We extended our prisma client with [our custom bignumber extension](https://www.prisma.io/docs/orm/prisma-client/client-extensions) that converts `string` fields into [`BigNumber`](https://www.npmjs.com/package/bignumber.js) object.
And now, generated types by `@pothos/prisma-plugin` generator is mismatching.

```typescript
Types of property `price` are incompatible. Type `BigNumber` is not assignable to type `string`
```

**Comment by 2coo** [comment]
For now, migrated to `builder.objectRef` from `builder.prismaObject`

```typescript
const db = new PrismaClient().$extends(BigNumberExtension);

type ExtendedProduct = Prisma.Payload<
  typeof db.product,
  "findUnique"
>["scalars"];

export const ProductObject =
  builder.objectRef <
  ExtendedProduct >
  "Product".implement({
    fields: (t) => ({
      id: t.expose("id", {
        type: "ObjectID",
      }),
      name: t.exposeString("name"),
      description: t.exposeString("description"),
      price: t.field({
        type: "BigNumber",
        resolve: (product) => product.price,
      }),
    }),
  });

builder.queryField("product", (t) =>
  t.field({
    type: ProductObject,
    args: {
      id: t.arg({
        type: "ObjectID",
        required: false,
      }),
    },
    resolve: async (_parent, { id }) => {
      // now there is no type mismatching error
      return db.product.findUniqueOrThrow({
        where: { id },
      });
    },
  })
);

```

**Comment by 2coo** [comment]
Instead of relying on `@pothos/prisma-plugin` generator. 
Can we infer all types from Prisma Client using [type utilities](https://www.prisma.io/docs/orm/prisma-client/type-safety#type-utilities) in the future pothos version?

**Reply by hayes** [reply, marked as answer]
It should be possible to entirely infer the generated types from the prisma client instead.  The first versions of the prisma plugin worked this way.  It ended up causing type-checking performance problems, and generating the types was the easiest way to get types that would be efficient for type-checking, but I don't think there is something that would prevent creating a type like: `PrismaTypesFromClient<typeof prisma>` that could be used instead of the generated types.

There are some relevant examples of extracting types from the client here: https://github.com/prisma/prisma/issues/5315#issuecomment-2059432225.  I'd be open adding a utility type for this to the plugin, but I'm not sure I have time right now to prioritize it.  If you are interested in contributing it, I'd be more than happy to help out and point you in the right direction!

**Reply by hayes** [reply]
This is the interface that needs to be generated for each model:

```ts

export interface PrismaModelTypes {
  Name: string;
  Shape: {};
  Include: unknown;
  Select: unknown;
  OrderBy: unknown;
  Where: {};
  WhereUnique: {};
  Create: {};
  Update: {};
  ListRelations: string;
  RelationName: string;
  Relations: Record<
    string,
    {
      Shape: unknown;
      Name: string;
      Nullable: boolean;
    }
  >;
}
```

**Reply by hayes** [reply]
`ListRelations` and `RelationName` would be string unions like `'author' | 'comments'`

**Comment by 2coo** [comment]
For some people faced the same problem as mine.

Here is [`PrismaTypesFromClient`](https://gist.github.com/2coo/11df1d175444e922571fab8c7e1ce6af) type that infers types from `PrismaClient`


Example:
```typescript
import SchemaBuilder from "@pothos/core"
import { BigNumberExtension } from "~/lib/prisma-extensions"
// @pothos/plugin-prisma@4.0.4
import type { PrismaTypesFromClient } from "@pothos/plugin-prisma"

const db = new PrismaClient().$extend(BigNumberExtension)

export const builder = new SchemaBuilder<{
  PrismaTypes: PrismaTypesFromClient<typeof db>
}>()
```

Created a [PR](https://github.com/hayes/pothos/pull/1230) that adds that utility type into `@pothos/prisma-plugin`
